rfc/text/wamp-terminology.md

```diff
@@ -0,0 +1,80 @@
+## Terminology {#terminology}
+
+This chapter collects the terminology *as used in WAMP*, the group of specialized words and respective meanings
+as used in this specification.
+
+Implementations SHOULD use terms as defined here in their public interfaces and documentation,
+and SHOULD NOT reinvent or reinterpret terms. Users SHOULD be able to transfer their knowledge gained on WAMP
+in the context of one implementation when starting to use another implementation.
+This is to support the overarching goal of WAMP to free application developers from restrictions when
+building distributed applications, both at the network level, and when choosing or switching the WAMP client
+libraries or routers used.
+
+Our goal is to maximize **User Choice and Experience** for developers of WAMP based applications,
+both formally (open protocol and open source) as well as practically (switching costs).
+
+
+### Fundamental
+
+{align="left"}
+| Term                              | Definition                                                                                              |
+|-----------------------------------|---------------------------------------------------------------------------------------------------------|
+| *User*                            | A user runs *Client*s or *Router*s which implement WAMP                                                 |
+| *Client*                          | A program run by a *User* with application code using WAMP for application level communication          |
+| *Router*                          | A program run by a *User* with middleware code using WAMP to provide application routing services       |
+| *Broker*                          | *FIXME*                                                                                                 |
+| *Dealer*                          | *FIXME*                                                                                                 |
+| *Peer*                            | A WAMP *Client* or *Router*. An implementation might embed, provide or use both roles                   |
+| *Realm*                           | Isolated WAMP URI namespace, routing and administrative domain, optionally protected by **AA**          |
+| *Transport*                       | A message-based, reliable, ordered, bidirectional (full-duplex) channel over which *Peers* communicate  |
+| *Connection*                      | An underlying entity (if any) carrying the *Transport*, e.g. a network connection, pipe, queue or such  |
+| *Session*                         | Transient conversation between a *Client* and a *Router* on a *Realm* over a *Transport*                |
+| *Message*                         | *FIXME*                                                                                                 |
+| *Serializer*                      | *FIXME*                                                                                                 |
```

Comment:
Interesting how to describe a serializer without saying that: `Serializer is just a serializer` :)

How about this: 

The Serializer is responsible for turning data objects and WAMP messages into a specific format convenient for transporting (JSON, CBOR, MsgPack, etc) and vice versa.

Reply:
A Serializer is a function from WAMP Messages to bytes - as bytes (a chunk of) is what is needed by WAMP Transports.

So Serializer, Message, Transport and Bytes (of a message as taken/consumed/transported by Transport) all relate

Reply:
actually, there are two meanings of "serializer" in WAMP. let me explain.

technically and in general (independent of WAMP), a "raw serializer" **RS** is

1. if you take **O** as the set of all possible values of objects in the host language, and
2.  **B** the set of all Bytes strings which encode possible values which are valid under **RS**

then **RS** is a function `RS: O -> B`.

couple of notes:
* of course no every Bytes string is "proper" under **RS** (there are Bytes string which are invalid, as hence **B** is a subset of the set of all Bytes string)
* of course **RS** should be "universal" with respect to the host language (every valid host language value should be serializable under **RS** .. otherwise some application payloads might not be possible to be sent or received in WAMP actions)

Moreover, what's interesting to note:

* it would be desirable if **RS** defined a *bijective* function  `RS: O -> B`
* however, the only serializer with that property I know of is [Bencode](https://en.wikipedia.org/wiki/Bencode)
* none of the currently defined WAMP serializers has this property

now, given **RS**, when specifically encoding WAMP messages with arbitrary application payloads. the application payloads are arbitrary (if the serializer is universal), but the fact that this is about WAMP messages means, the "outer" stuff is restricted by the WAMP spec (for WAMP messages). let's call that serializer **WS**. then

1. the set of all valid WAMP messages **W**, that is objects in the host language which are valid WAMP message with any application payload, and have arbitrary application payloads
2. the set of all Bytes strings **BW** which **WS** deserializes into messages in **W** with arbitrary application payloads, then

**WS** is a function `WS: W->BW`

Reply:
> however, the only serializer with that property I know of is [Bencode](https://en.wikipedia.org/wiki/Bencode)

I'm not sure, but [Deterministically-Encoded CBOR](https://www.rfc-editor.org/rfc/rfc8949.html#name-deterministically-encoded-c) may also have that property. Two recent implementations of such have been [announced on the CBOR mailing list](https://mailarchive.ietf.org/arch/msg/cbor/l7nzQHFjfpK9nfBOHiQ1L-Rr558/).

Reply:
Well, that is cool and maybe even interesting to some readers. But let's start with simple descriptions and then we can elaborate more on that.

Reply:
For "serializer", how about: "Pair of functions that convert a Message (or Message Payload) into bytes and vice-versa"

In the spec, the term "serializer" is used loosely to also include deserialization operations. In electronics, [SerDes](https://en.wikipedia.org/wiki/SerDes), refers to a block that performs both serialization and deserialization, but it is not commonly used in software.

_Codec_ (coder/decoder) is a term that is more commonly found in software, but most people think "audio/video" when they see that term.

I think most readers will understand from context that "serializer" also includes the facilities for converting bytes back into messages/payloads.